ls/src/file.rs

```diff
@@ -0,0 +1,302 @@
+
+use crate::flags::Flags;
+
+#[derive(PartialEq, Eq)]
+pub(crate) enum FileColor {
+    Show,
+    Hide,
+}
+
+pub(crate) type Files = Vec<File>;
+
+/// Represents a file and it's properties
+pub(crate) struct File {
+    pub name: BString,
+    pub path: PathBuf,
+    pub metadata: fs::Metadata,
+    flags: Flags,
+}
+
+impl File {
+    /// Creates a `File` instance from a `PathBuf`
+    pub fn from(path: PathBuf, flags: Flags) -> io::Result<Self> {
+        let metadata = File::metadata(&path, &flags)?;
+
+        if flags.dereference && metadata.file_type().is_symlink() {
+            let symlink = fs::read_link(path)?;
+
+            let name = File::path_buf_to_file_name(&symlink)?;
+
+            let metadata = File::metadata(&symlink, &flags)?;
+
+            return Ok(File { name, path: symlink, metadata, flags });
+        }
+
+        let name = File::path_buf_to_file_name(&path)?;
+
+        Ok(File { name, path, metadata, flags })
+    }
+
+    /// Creates a `File` instance from a `PathBuf` and supplies a file name
+    pub fn from_name(name: BString, path: PathBuf, flags: Flags) -> io::Result<Self> {
+        let metadata = File::metadata(&path, &flags)?;
+
+        Ok(File { name, path, metadata, flags })
+    }
+
+    /// Retrieves the number of blocks allocated to a file as a string
+    pub fn blocks(&self) -> u64 {
+        let blocks = self.metadata.blocks();
+
+        if self.flags.block_size {
+            let st_size = blocks * 512;
+
+            st_size / 1024
+        } else {
+            blocks
+        }
+    }
+
+    /// Retrieves a files permissions as a string
+    pub fn permissions(&self) -> String {
+        let mode = self.metadata.permissions().mode();
+
+        unix_mode::to_string(mode)
+    }
+
+    /// Retrieves the number of hard links pointing to a file as a string
+    pub fn hard_links(&self) -> String { self.metadata.nlink().to_string() }
+
+    /// Retrieves the inode number as a string
+    pub fn inode(&self) -> String { self.metadata.ino().to_string() }
+
+    /// Retrieves the file's user name as a string. If the `-n` flag is set,
+    /// the the user's ID is returned
+    pub fn user(&self) -> Result<BString, PasswdError> {
+        if self.flags.numeric_uid_gid {
+            return Ok(BString::from(self.metadata.uid().to_string()));
+        }
+
+        match Passwd::from_uid(self.metadata.uid()) {
+            Ok(passwd) => Ok(passwd.name().to_owned()),
+            Err(err) => Err(err),
+        }
+    }
+
+    /// Retrieves the file's group name as a string. If the `-n` flag is set,
+    /// the the group's ID is returned
+    pub fn group(&self) -> Result<BString, GroupError> {
+        if self.flags.numeric_uid_gid {
+            return Ok(BString::from(self.metadata.gid().to_string()));
+        }
+
+        match Group::from_gid(self.metadata.gid()) {
+            Ok(group) => Ok(group.name().to_owned()),
+            Err(err) => Err(err),
+        }
+    }
+
```

Comment:
In a recent commit I changed the `Group` and `Passwd` to return `io::Result`, because all os-module-like public API seems to return `io::Error` on problems, and most of the rest of the Core OS API was already returning a `io::Error`, so it made sense to change for better consistency with the Core API and with the Rust ecosystem.

Can you change that? I also can change that before merging, I have no problems with that